sklearn/ensemble/_hist_gradient_boosting/tests/test_loss.py

```diff
@@ -0,0 +1,192 @@
+    y_true = np.array([y_true], dtype=Y_DTYPE)
+    x0 = np.array([x0], dtype=Y_DTYPE).reshape(1, 1)
+    get_gradients, get_hessians = get_derivatives_helper(loss)
+
+    def func(x):
+        return loss(y_true, x)
+
+    def fprime(x):
+        return get_gradients(y_true, x)
+
+    def fprime2(x):
+        return get_hessians(y_true, x)
+
+    optimum = newton(func, x0=x0, fprime=fprime, fprime2=fprime2)
+    assert np.allclose(loss.inverse_link_function(optimum), y_true)
+    assert np.allclose(loss(y_true, optimum), 0)
+    assert np.allclose(get_gradients(y_true, optimum), 0)
+
+
+@pytest.mark.parametrize('loss, n_classes, prediction_dim', [
+    ('least_squares', 0, 1),
+    ('binary_crossentropy', 2, 1),
+    ('categorical_crossentropy', 3, 3),
+])
+@pytest.mark.skipif(Y_DTYPE != np.float64,
+                    reason='Need 64 bits float precision for numerical checks')
+def test_numerical_gradients(loss, n_classes, prediction_dim):
+    # Make sure gradients and hessians computed in the loss are correct, by
+    # comparing with their approximations computed with finite central
+    # differences.
+    # See https://en.wikipedia.org/wiki/Finite_difference.
+
+    rng = np.random.RandomState(0)
+    n_samples = 100
+    if loss == 'least_squares':
+        y_true = rng.normal(size=n_samples).astype(Y_DTYPE)
+    else:
+        y_true = rng.randint(0, n_classes, size=n_samples).astype(Y_DTYPE)
+    raw_predictions = rng.normal(
+        size=(prediction_dim, n_samples)
+    ).astype(Y_DTYPE)
+    loss = _LOSSES[loss]()
+    get_gradients, get_hessians = get_derivatives_helper(loss)
+
+    # only take gradients and hessians of first tree / class.
+    gradients = get_gradients(y_true, raw_predictions)[0, :].ravel()
+    hessians = get_hessians(y_true, raw_predictions)[0, :].ravel()
+
+    # Approximate gradients
+    # For multiclass loss, we should only change the predictions of one tree
+    # (here the first), hence the use of offset[:, 0] += eps
+    # As a softmax is computed, offsetting the whole array by a constant would
+    # have no effect on the probabilities, and thus on the loss
+    eps = 1e-9
+    offset = np.zeros_like(raw_predictions)
+    offset[0, :] = eps
+    f_plus_eps = loss(y_true, raw_predictions + offset / 2, average=False)
+    f_minus_eps = loss(y_true, raw_predictions - offset / 2, average=False)
+    numerical_gradient = (f_plus_eps - f_minus_eps) / eps
+
+    # Approximate hessians
+    eps = 1e-4  # need big enough eps as we divide by its square
+    offset[0, :] = eps
+    f_plus_eps = loss(y_true, raw_predictions + offset, average=False)
+    f_minus_eps = loss(y_true, raw_predictions - offset, average=False)
+    f = loss(y_true, raw_predictions, average=False)
+    numerical_hessians = (f_plus_eps + f_minus_eps - 2 * f) / eps**2
+
+    def relative_error(a, b):
+        return np.abs(a - b) / np.maximum(np.abs(a), np.abs(b))
+
+    assert np.all(relative_error(numerical_gradient, gradients) < 1e-5)
```

Comment:
Why do we need to have a symmetric version of `assert_allclose`? Isn't it enough to keep the desired to compute the relative error.